### ISSUE TITLE
Incorrect type definitions for single row queries

### ISSUE THREAD

**Issue by spacejack** [tgriesser/knex#3229]
Knex version: 0.17.0

Example:

```typescript
const result = await knex('foo').first('id')
```

`result` then has the type `any[]` which breaks previously working code. Eg:

```typescript
const {id} = await knex('foo').first('id')
```
Now results in the error: `Property 'id' does not exist on type 'any[]'`

The following also does not work:

```typescript
const result: {id: number} = await knex('foo').first('id')
```

As a typescript user I appreciate continued efforts to improve types but I feel that unless a distinction can be made between single/multi-row selects then result types should revert back to `any`.

**Comment by kibertoad** [comment]
@lorefnon

**Comment by spacejack** [comment]
Wow, I see you've already tackled the issue. Thanks for the quick response!

I should have added that it would be fair to return a type of `T | undefined` for a single row query. But I'm not sure how precise of a type you're implementing. `any` is fine as well.

**Comment by kibertoad** [comment]
@lorefnon Any reason why we might prefer `any` over `T | undefined?`

**Comment by spacejack** [comment]
No, `T | undefined` is preferable. I'm not sure how to get a type for `T` though?

**Comment by lorefnon** [comment]
Yes, you are right. This should include undefined.

**Comment by kibertoad** [comment]
@spacejack Could you please try knex@0.17.1-next and see if it solves your problem in satisfactory way?

**Comment by spacejack** [comment]
@kibertoad thanks for the update. With this version, the example I showed above now returns an 'any' type, so I can apply my own type as before:
```typescript
const result: {id: number} | undefined = await knex('foo').first('id')
```
Also works with 'where' clauses:
```typescript
const result: {id: number} | undefined = await knex('foo').first('id').where({bar})
```
However if I chain a `count()`, `sum()` or other aggregate functions then the result type switches back to an array. Unfortunately I use `first()` a lot when doing aggregates.
```typescript
const result: {count: number} = await knex('foo').first('id').count({count: '*'})
```

**Comment by kibertoad** [comment]
@spacejack 0.17.1 with fixes for this issue is out!

**Comment by spacejack** [comment]
Thank you, it seems more correct now for single-row queries.

It is still a bit awkward to declare the response types for both single and multi-row queries however. These no longer work:
```typescript
const result: {count: number} = await knex('foo').first().count({count: '*'})
const results: {count: number}[] = await knex('foo').count({count: '*'})
```
`Property 'count' is missing in type 'Dict<string | number>' but required in type '{ count: number; }'`
I can do this:
```typescript
const {count} = await knex('foo').first().count({count: '*'})
```
But count is `string | number`.

The following works but gets a bit ugly with the parentheses, especially in chained queries:
```typescript
const {count} = (await knex('foo').first().count({count: '*'})) as {count: number}
```
Alternately:
```typescript
const {count} = await knex('foo').first().count({count: '*'})
const c = count as number
```
I'm not sure what the least boilerplatey way is to declare response types.

**Comment by lorefnon** [comment]
I believe your original usage should continue to work, and after https://github.com/tgriesser/knex/pull/3239 it will. 

Result of `await knex('foo').first().count({count: '*'});` will be inferred as `{count: any}` and so it will be assignable to `{count; number}`.

For more control it would be possible to pass a type param to count: 

```
await knex('foo').first().count<{count: number}>({count: '*'});
```

**Comment by kibertoad** [comment]
Fix released in 0.17.2.

**Comment by spacejack** [comment]
Thanks! Here is another case:
```ts
const rows = await knex('session')
    .select({userId : 'userId', duration: knex.raw('max(endedAt) - min(startedAt)')})
    .countDistinct({days: knex.raw(
        `strftime('%Y%m%d', datetime(startedAt/1000, 'unixepoch', 'localtime'))`
    )})
```
The result type is `{days: any}[]` and does not include the `userId` or `duration` fields. Same if using `first()` instead of `select()` (though it is a single object not array.)

It is solvable as described in my last comment by appending `as {userId: number, duration: number, days: number}[]`

**Comment by kibertoad** [comment]
@spacejack Fixed in 0.17.3 thanks to tireless ninja @lorefnon!

**Comment by ravishivt** [comment]
There's still an issue with `first('*')` as it still returns `any[]`.  Workaround is to not specify any columns which defaults to `*`, i.e. `first()`.
